File: reference/typescript-src/agents/impl/market-structure-agent.ts
// ============================================================================
// AGENT 1: MARKET STRUCTURE AGENT
// Detects BOS, CHOCH, MSS, IDM, Internal/External Structure
// ============================================================================

import { TradingAgent, AgentContext, AgentOutput, AgentInsight, AgentName } from '../types';
import { Bias, Direction } from '../../core/types';
import { MarketStructureAnalyzer } from '../../modules/market-structure';

let insightSeq = 0;

export class MarketStructureAgent implements TradingAgent {
  readonly name: AgentName = 'MARKET_STRUCTURE';
  readonly description = 'Detects BOS, CHOCH, MSS, IDM, internal/external structure';
  readonly version = '1.0.0';
  private analyzer = new MarketStructureAnalyzer();

  analyze(context: AgentContext): AgentOutput {
    const start = performance.now();
    const { candles } = context;
    const insights: AgentInsight[] = [];

    if (candles.length < 20) return this.emptyOutput(start);

    const structure = this.analyzer.analyze(candles);

    // Generate insights from recent structure breaks (last 5)
    const recent = structure.structureBreaks.slice(-5);
    for (const brk of recent) {
      const weight = brk.type === 'MSS' ? 3 : brk.type === 'CHOCH' ? 2.5 : brk.type === 'BOS' ? 2 : 1;
      insights.push({
        id: `ms_${++insightSeq}`,
        agentName: this.name,
        type: brk.type,
        direction: brk.direction,
        confidence: brk.confirmed ? 85 : 60,
        price: brk.breakPrice,
        timestamp: brk.breakTimestamp,
        barIndex: brk.breakIndex,
        detail: `${brk.type} ${brk.direction} at ${brk.breakPrice.toFixed(5)} (${brk.structureType})`,
        weight,
        tags: [brk.structureType, brk.confirmed ? 'CONFIRMED' : 'UNCONFIRMED'],
      });
    }

    // Internal vs External structure insights
    const internalCount = structure.internalStructure.length;
    const externalCount = structure.externalStructure.length;
    if (externalCount > 0) {
      const lastExt = structure.externalStructure[structure.externalStructure.length - 1];
      insights.push({
        id: `ms_${++insightSeq}`, agentName: this.name, type: 'EXTERNAL_STRUCTURE',
        direction: lastExt.direction, confidence: 80, price: lastExt.breakPrice,
        timestamp: lastExt.breakTimestamp, detail: `External structure: ${lastExt.type} ${lastExt.direction}`,
        weight: 2.5, tags: ['EXTERNAL'],
      });
    }

    const bias = structure.currentBias;
    const confidence = this.calculateConfidence(structure.structureBreaks.slice(-5), bias);

    const narrative = `Structure bias: ${bias}. Recent breaks: ${recent.map(b => `${b.type} ${b.direction}`).join(', ')}.` +
      ` ${internalCount} internal, ${externalCount} external breaks total.`;

    return {
      agentName: this.name, status: 'COMPLETE', bias, confidence, insights,
      narrative, processingTimeMs: performance.now() - start, timestamp: Date.now(),
    };
  }

  private calculateConfidence(recent: any[], bias: Bias): number {
    if (bias === 'NEUTRAL') return 40;
    const aligned = recent.filter((b: any) => b.direction === bias).length;
    return Math.min(95, 50 + aligned * 10);
  }

  private emptyOutput(start: number): AgentOutput {
    return {
      agentName: this.name, status: 'COMPLETE', bias: 'NEUTRAL', confidence: 30,
      insights: [], narrative: 'Insufficient data for structure analysis.',
      processingTimeMs: performance.now() - start, timestamp: Date.now(),
    };
  }

  reset(): void { this.analyzer.reset(); }
}

File: app/src/main/java/com/foxtrader/app/domain/usecase/ai/agents/MarketStructureAgent.kt
package com.foxtrader.app.domain.usecase.ai.agents

import com.foxtrader.app.domain.model.AgentContext
import com.foxtrader.app.domain.model.AgentInsight
import com.foxtrader.app.domain.model.AgentName
import com.foxtrader.app.domain.model.AgentOutput
import com.foxtrader.app.domain.model.AgentStatus
import com.foxtrader.app.domain.model.StructureBreakType
import com.foxtrader.app.domain.usecase.AnalyzeMarketStructureUseCase
import com.foxtrader.app.domain.usecase.ai.TradingAgent
import javax.inject.Inject

/**
 * MARKET_STRUCTURE agent — reads swing structure to derive HTF directional bias
 * and emits BOS / CHoCH / MSS insights. Foundational agent (runs in phase 1).
 */
class MarketStructureAgent @Inject constructor(
    private val analyzeStructure: AnalyzeMarketStructureUseCase,
) : TradingAgent {

    override val name = AgentName.MARKET_STRUCTURE
    override val description = "Detects BOS/CHoCH/MSS and derives HTF directional bias."
    override val version = "1.0.0"

    override fun analyze(context: AgentContext): AgentOutput {
        val start = System.nanoTime()
        val candles = context.candles
        if (candles.size < 20) {
            return neutralOutput(name, "Insufficient data for structure analysis.", start)
        }

        val structure = analyzeStructure(candles)
        val recent = structure.breaks.filter { it.confirmed }.takeLast(5)
        if (recent.isEmpty()) {
            return neutralOutput(name, "No confirmed structure breaks yet.", start)
        }

        val insights = recent.map { brk ->
            AgentInsight(
                id = "${name}-${brk.type}-${brk.breakIndex}",
                agentName = name,
                type = brk.type.name, // BOS / CHOCH / MSS / IDM
                direction = brk.direction,
                confidence = weightFor(brk.type) * 60.0, // scaled per break type
                price = brk.breakPrice,
                timestamp = brk.breakTimestamp,
                barIndex = brk.breakIndex,
                detail = "${brk.type} ${brk.direction} @ ${brk.breakPrice}",
                weight = weightFor(brk.type),
                tags = listOf(brk.type.name),
            )
        }

        val bias = structure.bias
        val aligned = recent.count { bias.agreesWith(it.direction) }
        val confidence = (aligned.toDouble() / recent.size) * 100.0

        return AgentOutput(
            agentName = name,
            status = AgentStatus.COMPLETE,
            bias = bias,
            confidence = confidence,
            insights = insights,
            narrative = "Structure bias $bias from ${recent.size} recent breaks " +
                "($aligned aligned). Latest: ${recent.last().type} ${recent.last().direction}.",
            processingTimeMs = elapsedMs(start),
            timestamp = System.currentTimeMillis(),
        )
    }

    private fun weightFor(type: StructureBreakType): Double = when (type) {
        StructureBreakType.MSS -> 1.5
        StructureBreakType.CHOCH -> 1.3
        StructureBreakType.BOS -> 1.0
        StructureBreakType.IDM -> 0.6
    }
}
